File: backend/app/api/personal_secrets.py
from typing import Annotated

import oracledb
from fastapi import APIRouter, Depends, HTTPException, Path, Query, status

from app.api.errors import oracle_http_exception
from app.core.security import require_current_user
from app.repositories.personal_secrets import (
    create_personal_secret,
    delete_personal_secret,
    get_personal_secret,
    list_personal_secrets,
    reveal_personal_secret,
    update_personal_secret,
)
from app.repositories.users import AuthContext
from app.schemas.personal_secrets import (
    PersonalSecretCreate,
    PersonalSecretItem,
    PersonalSecretListResponse,
    PersonalSecretRevealRequest,
    PersonalSecretRevealResponse,
    PersonalSecretUpdate,
)
from app.services.personal_secret_crypto import PersonalSecretCryptoError


router = APIRouter(prefix="/personal-secrets", tags=["personal-secrets"])


@router.get("", response_model=PersonalSecretListResponse)
async def get_personal_secrets(
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
    include_total: bool = True,
    q: Annotated[str | None, Query(min_length=1, max_length=400)] = None,
    auth_context: AuthContext = Depends(require_current_user),
) -> PersonalSecretListResponse:
    try:
        items, total = await list_personal_secrets(
            limit=limit,
            offset=offset,
            include_total=include_total,
            q=q,
            auth_context=auth_context,
        )
    except PersonalSecretCryptoError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
    except oracledb.Error as exc:
        raise oracle_http_exception(exc, "Oracle rejected the personal secret query") from exc

    return PersonalSecretListResponse(items=items, total=total, limit=limit, offset=offset)


@router.post("", response_model=PersonalSecretItem, status_code=status.HTTP_201_CREATED)
async def post_personal_secret(
    payload: PersonalSecretCreate,
    auth_context: AuthContext = Depends(require_current_user),
) -> PersonalSecretItem:
    try:
        created = await create_personal_secret(payload, auth_context)
    except PersonalSecretCryptoError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
    except oracledb.Error as exc:
        raise oracle_http_exception(exc, "Oracle rejected the personal secret entry") from exc

    return PersonalSecretItem.model_validate(created)


@router.get("/{secret_id}", response_model=PersonalSecretItem)
async def get_personal_secret_detail(
    secret_id: Annotated[int, Path(ge=1)],
    auth_context: AuthContext = Depends(require_current_user),
) -> PersonalSecretItem:
    try:
        item = await get_personal_secret(secret_id, auth_context)
    except PersonalSecretCryptoError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
    except oracledb.Error as exc:
        raise oracle_http_exception(exc, "Oracle rejected the personal secret detail query") from exc

    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Personal secret not found")
    return PersonalSecretItem.model_validate(item)


@router.patch("/{secret_id}", response_model=PersonalSecretItem)
async def patch_personal_secret(
    secret_id: Annotated[int, Path(ge=1)],
    payload: PersonalSecretUpdate,
    auth_context: AuthContext = Depends(require_current_user),
) -> PersonalSecretItem:
    try:
        item = await update_personal_secret(secret_id, payload, auth_context)
    except PersonalSecretCryptoError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
    except oracledb.Error as exc:
        raise oracle_http_exception(exc, "Oracle rejected the personal secret update") from exc

    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Personal secret not found")
    return PersonalSecretItem.model_validate(item)


@router.delete("/{secret_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_personal_secret_item(
    secret_id: Annotated[int, Path(ge=1)],
    auth_context: AuthContext = Depends(require_current_user),
) -> None:
    try:
        deleted = await delete_personal_secret(secret_id, auth_context)
    except oracledb.Error as exc:
        raise oracle_http_exception(exc, "Oracle rejected the personal secret delete") from exc

    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Personal secret not found")


@router.post("/{secret_id}/reveal", response_model=PersonalSecretRevealResponse)
async def post_personal_secret_reveal(
    secret_id: Annotated[int, Path(ge=1)],
    payload: PersonalSecretRevealRequest,
    auth_context: AuthContext = Depends(require_current_user),
) -> PersonalSecretRevealResponse:
    try:
        value = await reveal_personal_secret(secret_id, payload.field, auth_context)
    except PersonalSecretCryptoError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
    except oracledb.Error as exc:
        raise oracle_http_exception(exc, "Oracle rejected the personal secret reveal") from exc

    if value is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Personal secret not found")
    if payload.field == "all":
        return PersonalSecretRevealResponse(field=payload.field, values=value)
    return PersonalSecretRevealResponse(field=payload.field, value=value.get(payload.field))

File: frontend/src/api/personalSecrets.ts
import type { PersonalSecretDraft, PersonalSecretItem, PersonalSecretRevealField } from "../types";
import { buildQuery, readCachedGet, request } from "./client";

export interface PersonalSecretListResponse {
  items: PersonalSecretItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface PersonalSecretQuery {
  query?: string;
  limit: number;
  offset: number;
  includeTotal?: boolean;
}

export interface PersonalSecretRevealResponse {
  field: PersonalSecretRevealField;
  value?: string | null;
  values?: Record<string, string | null> | null;
}

export async function fetchPersonalSecrets(query: PersonalSecretQuery): Promise<PersonalSecretListResponse> {
  return request<PersonalSecretListResponse>(buildPersonalSecretsPath(query));
}

export function readCachedPersonalSecrets(query: PersonalSecretQuery): PersonalSecretListResponse | null {
  return readCachedGet<PersonalSecretListResponse>(buildPersonalSecretsPath(query), 15 * 60 * 1000);
}

export async function getPersonalSecret(id: number): Promise<PersonalSecretItem> {
  return request<PersonalSecretItem>(`/api/personal-secrets/${id}`);
}

export async function createPersonalSecret(draft: PersonalSecretDraft): Promise<PersonalSecretItem> {
  return request<PersonalSecretItem>("/api/personal-secrets", {
    method: "POST",
    invalidatePrefixes: ["/api/personal-secrets"],
    body: JSON.stringify(toPayload(draft)),
  });
}

export async function updatePersonalSecret(id: number, draft: PersonalSecretDraft): Promise<PersonalSecretItem> {
  return request<PersonalSecretItem>(`/api/personal-secrets/${id}`, {
    method: "PATCH",
    invalidatePrefixes: ["/api/personal-secrets"],
    body: JSON.stringify(toPayload(draft)),
  });
}

export async function deletePersonalSecret(id: number): Promise<void> {
  return request<void>(`/api/personal-secrets/${id}`, {
    method: "DELETE",
    invalidatePrefixes: ["/api/personal-secrets"],
  });
}

export async function revealPersonalSecret(id: number, field: PersonalSecretRevealField): Promise<PersonalSecretRevealResponse> {
  return request<PersonalSecretRevealResponse>(`/api/personal-secrets/${id}/reveal`, {
    method: "POST",
    body: JSON.stringify({ field }),
  });
}

function buildPersonalSecretsPath(query: PersonalSecretQuery): string {
  return `/api/personal-secrets${buildQuery({
    limit: String(query.limit),
    offset: String(query.offset),
    q: query.query,
    include_total: query.includeTotal === false ? false : undefined,
  })}`;
}

function toPayload(draft: PersonalSecretDraft) {
  return {
    system_name: draft.system_name,
    login_url: draft.login_url || null,
    username: draft.username || null,
    password: draft.password || null,
    notes: draft.notes || null,
    tags: draft.tags || null,
  };
}
